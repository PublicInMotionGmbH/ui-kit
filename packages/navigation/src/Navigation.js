import React from 'react'
import PropTypes from 'prop-types'
import cls from 'classnames'

import { prefix } from '@talixo/shared'

const moduleName = prefix('navigation')

/**
 * Component which represents Navigation.
 *
 * @param {object} props
 * @param {*} [props.children]
 * @param {string} [props.className]
 * @param {*} [props.divider]
 * @param {string} [props.type]
 * @returns {React.Element}
 */
function Navigation (props) {
  const { children, className, divider, type, ...passedProps } = props

  /**
   * Method that return object with passed props.
   *
   * @param {array} children
   * @param {number} i
   * @returns {object}
   */
  const buildChildProps = (children, i) => {
    const typeClassName = `${moduleName}_element--${type}`
    const newProps = { typeClassName }

    if (type === 'breadcrumbs') {
      return { ...newProps, divider: children.length <= 1 ? divider : i > children.length - 2 ? null : divider }
    }

    return newProps
  }

  const mappedChildren = React.Children.map(children, (child, i) => {
    return React.cloneElement(child, buildChildProps(children, i))
  })

  return (
    <ul className={cls(className, moduleName, `${moduleName}--${type}`)} {...passedProps}>
      {mappedChildren}
    </ul>
  )
}

Navigation.propTypes = {
  /** Navigation items */
  children: PropTypes.node,

  /** Additional class name */
  className: PropTypes.string,

  /** Divider */
  divider: PropTypes.node,

  /** Type of navigation */
  type: PropTypes.oneOf(['navigation', 'pagination', 'breadcrumbs', 'tabs'])
}

Navigation.defaultProps = {
  divider: '/',
  type: 'navigation'
}

export default Navigation
