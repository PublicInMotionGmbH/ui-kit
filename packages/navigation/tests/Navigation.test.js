import React from 'react'
import { shallow } from 'enzyme'

import Navigation, { moduleName } from '../src/Navigation'

const includesClassName = (wrapper, className) => {
  const clsName = wrapper.props().className
  expect(clsName.includes(className)).toEqual(true)
}

describe('Module name', () => {
  it('is passed correctly', () => {
    const wrapper = shallow(<Navigation />)

    includesClassName(wrapper, moduleName)
  })
})

describe('<Navigation />', () => {
  const customDivider = '#'

  it('renders children correctly', () => {
    const wrapper = shallow(
      <Navigation>
        <span>Home</span>
        <span>Issues</span>
        <span>Major</span>
      </Navigation>
    )

    expect(wrapper.props().className.includes('navigation')).toEqual(true)
  })

  it('renders types correctly', () => {
    const wrappers = {
      navigation: shallow(<Navigation type='navigation' />),
      pagination: shallow(<Navigation type='pagination' />),
      breadcrumbs: shallow(<Navigation type='breadcrumbs' />),
      tabs: shallow(<Navigation type='tabs' />),
      steps: shallow(<Navigation type='steps' />)
    }

    Object.keys(wrappers).forEach(key => {
      includesClassName(wrappers[key], key)
    })
  })

  it('inserts dividers correctly', () => {
    const wrappers = {
      empty: shallow(<Navigation divider={customDivider} />),
      single: shallow(
        <Navigation divider={customDivider}>
          <span>Home</span>
        </Navigation>
      ),
      multiple: shallow(
        <Navigation divider={customDivider}>
          <span>Home</span>
          <span>Issues</span>
        </Navigation>
      )
    }

    expect(wrappers.empty.contains(customDivider)).toEqual(false)
    expect(wrappers.single.contains(customDivider)).toEqual(false)
    expect(wrappers.multiple.contains(customDivider)).toEqual(true)
  })

  it('inserts dividers between children elements', () => {
    const wrapper = shallow(
      <Navigation divider={customDivider}>
        <span>1</span>
        <span>2</span>
        <span>3</span>
      </Navigation>
    )

    expect(wrapper.childAt(0).contains(customDivider)).toEqual(false)
    expect(wrapper.childAt(1).contains(customDivider)).toEqual(true)
    expect(wrapper.childAt(2).contains(customDivider)).toEqual(false)
    expect(wrapper.childAt(3).contains(customDivider)).toEqual(true)
    expect(wrapper.childAt(4).contains(customDivider)).toEqual(false)
  })
})
