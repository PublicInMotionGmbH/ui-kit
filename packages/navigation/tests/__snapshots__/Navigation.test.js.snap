// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> rendering should render properly when children are passed 1`] = `
<NavigationWrapper
  parent={true}
  type="navbar"
>
  <NavigationElement
    render={[Function]}
  >
    test-element-0
  </NavigationElement>
  <NavigationElement
    render={[Function]}
  >
    test-element-1
  </NavigationElement>
  <NavigationElement
    render={[Function]}
  >
    test-element-2
  </NavigationElement>
</NavigationWrapper>
`;

exports[`<Navigation /> rendering should render properly when elements are passed via props 1`] = `
<NavigationWrapper
  parent={true}
  type="navbar"
>
  <NavigationElement
    disabled={false}
    id={0}
    key="0"
    label="Element 1"
    onClick={[Function]}
    render={[Function]}
    subelement={
      <NavigationWrapper
        panel={undefined}
        subtitle="Subtitile of element 1"
        type="navbar"
      >
        <NavigationElement
          disabled={false}
          id={10}
          label="Element 1 lvl 1"
          onClick={[Function]}
          render={[Function]}
          subelement={
            <NavigationWrapper
              panel={undefined}
              subtitle="Subtitile of element 1 lvl 1"
              type="navbar"
            >
              <NavigationElement
                disabled={false}
                id={20}
                label="Element 1 lvl 2"
                onClick={[Function]}
                panel={true}
                render={[Function]}
                subelement={
                  <NavigationWrapper
                    panel={true}
                    subtitle="Subtitile of element 1 lvl 2"
                    type="navbar"
                  >
                    <NavigationElement
                      disabled={false}
                      id={30}
                      label="Element 1 lvl 3"
                      onClick={[Function]}
                      render={[Function]}
                      subelement={null}
                      subtitle="Subtitile of element 1 lvl 3"
                      type="navbar"
                    />
                  </NavigationWrapper>
                }
                subelements={
                  Array [
                    Object {
                      "disabled": false,
                      "id": 30,
                      "label": "Element 1 lvl 3",
                      "onClick": [Function],
                      "subtitle": "Subtitile of element 1 lvl 3",
                    },
                  ]
                }
                subtitle="Subtitile of element 1 lvl 2"
                type="navbar"
              />
              <NavigationElement
                disabled={false}
                id={21}
                label="Element 2 lvl 2"
                onClick={[Function]}
                render={[Function]}
                subelement={null}
                subtitle="Subtitile of element 2 lvl 2"
                type="navbar"
              />
              <NavigationElement
                disabled={false}
                id={22}
                label="Element 3 lvl 2"
                onClick={[Function]}
                render={[Function]}
                subelement={null}
                subtitle="Subtitile of element 3 lvl 2"
                type="navbar"
              />
            </NavigationWrapper>
          }
          subelements={
            Array [
              Object {
                "disabled": false,
                "id": 20,
                "label": "Element 1 lvl 2",
                "onClick": [Function],
                "panel": true,
                "subelements": Array [
                  Object {
                    "disabled": false,
                    "id": 30,
                    "label": "Element 1 lvl 3",
                    "onClick": [Function],
                    "subtitle": "Subtitile of element 1 lvl 3",
                  },
                ],
                "subtitle": "Subtitile of element 1 lvl 2",
              },
              Object {
                "disabled": false,
                "id": 21,
                "label": "Element 2 lvl 2",
                "onClick": [Function],
                "subtitle": "Subtitile of element 2 lvl 2",
              },
              Object {
                "disabled": false,
                "id": 22,
                "label": "Element 3 lvl 2",
                "onClick": [Function],
                "subtitle": "Subtitile of element 3 lvl 2",
              },
            ]
          }
          subtitle="Subtitile of element 1 lvl 1"
          type="navbar"
        />
        <NavigationElement
          disabled={false}
          id={11}
          label="Element 2 lvl 1"
          onClick={[Function]}
          render={[Function]}
          subelement={null}
          subtitle="Subtitile of element 2 lvl 1"
          type="navbar"
        />
        <NavigationElement
          disabled={false}
          id={12}
          label="Element 3 lvl 1"
          onClick={[Function]}
          render={[Function]}
          subelement={null}
          subtitle="Subtitile of element 3 lvl 1"
          type="navbar"
        />
        <NavigationElement
          disabled={false}
          id={13}
          label="Element 4 lvl 1"
          onClick={[Function]}
          render={[Function]}
          subelement={null}
          subtitle="Subtitile of element 4 lvl 1"
          type="navbar"
        />
      </NavigationWrapper>
    }
    subelements={
      Array [
        Object {
          "disabled": false,
          "id": 10,
          "label": "Element 1 lvl 1",
          "onClick": [Function],
          "subelements": Array [
            Object {
              "disabled": false,
              "id": 20,
              "label": "Element 1 lvl 2",
              "onClick": [Function],
              "panel": true,
              "subelements": Array [
                Object {
                  "disabled": false,
                  "id": 30,
                  "label": "Element 1 lvl 3",
                  "onClick": [Function],
                  "subtitle": "Subtitile of element 1 lvl 3",
                },
              ],
              "subtitle": "Subtitile of element 1 lvl 2",
            },
            Object {
              "disabled": false,
              "id": 21,
              "label": "Element 2 lvl 2",
              "onClick": [Function],
              "subtitle": "Subtitile of element 2 lvl 2",
            },
            Object {
              "disabled": false,
              "id": 22,
              "label": "Element 3 lvl 2",
              "onClick": [Function],
              "subtitle": "Subtitile of element 3 lvl 2",
            },
          ],
          "subtitle": "Subtitile of element 1 lvl 1",
        },
        Object {
          "disabled": false,
          "id": 11,
          "label": "Element 2 lvl 1",
          "onClick": [Function],
          "subtitle": "Subtitile of element 2 lvl 1",
        },
        Object {
          "disabled": false,
          "id": 12,
          "label": "Element 3 lvl 1",
          "onClick": [Function],
          "subtitle": "Subtitile of element 3 lvl 1",
        },
        Object {
          "disabled": false,
          "id": 13,
          "label": "Element 4 lvl 1",
          "onClick": [Function],
          "subtitle": "Subtitile of element 4 lvl 1",
        },
      ]
    }
    subtitle="Subtitile of element 1"
    type="navbar"
  />
  <NavigationElement
    disabled={false}
    id={1}
    key="1"
    label="Element 2"
    onClick={[Function]}
    render={[Function]}
    subelement={null}
    subtitle="Subtitile of element 2"
    type="navbar"
  />
  <NavigationElement
    disabled={false}
    id={2}
    key="2"
    label="Element 3"
    onClick={[Function]}
    render={[Function]}
    subelement={null}
    subtitle="Subtitile of element 3"
    type="navbar"
  />
  <NavigationElement
    disabled={true}
    id={3}
    key="3"
    label="Element 4"
    onClick={[Function]}
    panel={true}
    render={[Function]}
    subelement={null}
    subtitle="Subtitile of element 4"
    type="navbar"
  />
</NavigationWrapper>
`;
